VideoMediaSampleRenderer can resolve its `WhenHasAvailableVideoFrame` callback with a flushed frame.

jyavenard opened the pull request:
#### 284f68f80bbb6d1f192c62f2b33acf0268232657
<pre>
VideoMediaSampleRenderer can resolve its `WhenHasAvailableVideoFrame` callback with a flushed frame.
<a href="https://bugs.webkit.org/show_bug.cgi?id=283443">https://bugs.webkit.org/show_bug.cgi?id=283443</a>
<a href="https://rdar.apple.com/140301261">rdar://140301261</a>

Reviewed by Jer Noble and Youenn Fablet.

If we decoded the video frame while we were in the middle of a flush, we would continue to notify the client
that a frame was available, even though that frame is flushed.

When we queue the task to notify the client that a frame has been decoded, we now pass the current flushId value, if it changes when the task runs, drop the message.

We also add some minimal logging.

Added tests.

* LayoutTests/media/media-rvfc-paused-mp4-expected.txt: Added.
* LayoutTests/media/media-rvfc-paused-mp4.html: Added.
* LayoutTests/media/media-rvfc-paused-offscreen-mp4-expected.txt: Added.
* LayoutTests/media/media-rvfc-paused-offscreen-mp4.html: Added.
* LayoutTests/media/media-rvfc-paused-offscreen-webm-expected.txt: Added.
* LayoutTests/media/media-rvfc-paused-offscreen-webm.html: Added.
* LayoutTests/media/media-rvfc-paused-webm-expected.txt: Added.
* LayoutTests/media/media-rvfc-paused-webm.html: Added.
* LayoutTests/media/media-vp8-hiddenframes.html: Remove FIXME and now use rVFC promise as the issue preventing its used was fixed in 286853@main
* LayoutTests/media/utilities.js:
(waitForVideoFrame): Promise was incorrectly resolved with two elements. Can only have one, return an array.
(waitForVideoFrameUntil): Ditto.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm: FlyBy: only perform a readback if we need to and not whenever a frame is available.
(WebCore::MediaPlayerPrivateWebM::setHasAvailableVideoFrame):
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::decodeNextSample):
(WebCore::VideoMediaSampleRenderer::decodedFrameAvailable):
(WebCore::VideoMediaSampleRenderer::maybeQueueFrameForDisplay):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::flush): Cancel timer before purging the compressed queue, to reduce the chance of having a midflight display.
(WebCore::VideoMediaSampleRenderer::notifyHasAvailableVideoFrame):

Canonical link: <a href="https://commits.webkit.org/286949@main">https://commits.webkit.org/286949@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ac3afb9269ad99dd9af73dd5dcd4edc186c9af66

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77669 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56702 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30591 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82296 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28947 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79786 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65858 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5002 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/60859 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18826 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80735 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50819 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66646 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41133 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/48209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24146 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27278 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/69323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24490 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83672 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5049 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3425 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69083 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5206 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66618 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68335 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17073 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12346 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10446 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4996 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/5011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/8449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6774 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -380,21 +380,24 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
 }
 
-void VideoMediaSampleRenderer::decodedFrameAvailable(RetainPtr<CMSampleBufferRef>&& sample)
+VideoMediaSampleRenderer::DecodedFrameResult VideoMediaSampleRenderer::decodedFrameAvailable(RetainPtr<CMSampleBufferRef>&& sample)
```
Why RetainPtr<>&&, it does not seem like you make use of it.

jyavenard replied:
we are moving it to the queue. existing code

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -380,21 +380,24 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
 
     purgeDecodedSampleQueue();
+
+    return result;
```
Why not returning a boolean telling whether to call `notifyHasAvailableVideoFrame`?

jernoble replied:
It looks like because it's necessary to distinguish between the "Too Early" and "Too Late" cases.

jyavenard replied:
> Why not returning a boolean telling whether to call `notifyHasAvailableVideoFrame`?

I also like that it documents what happened. So I could remove the comment.

jernoble commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
```diff
@@ -112,7 +117,7 @@ class VideoMediaSampleRenderer : public ThreadSafeRefCountedAndCanMakeThreadSafe
     bool shouldDecodeSample(CMSampleBufferRef, bool displaying);
 
-    void notifyHasAvailableVideoFrame();
+    void notifyHasAvailableVideoFrame(int);
```
I realize `std::atomic<int> m_flushId` already exists, but I kind of wish we had done:
```
using FlushID = int;
```

jyavenard replied:
I can change that.

jyavenard replied:
Excellent suggestion, makes the code more self-explanatory.

jyavenard committed:
VideoMediaSampleRenderer can resolve its `WhenHasAvailableVideoFrame` callback with a flushed frame.
https://bugs.webkit.org/show_bug.cgi?id=283443
rdar://140301261

Reviewed by Jer Noble and Youenn Fablet.

If we decoded the video frame while we were in the middle of a flush, we would continue to notify the client
that a frame was available, even though that frame is flushed.

When we queue the task to notify the client that a frame has been decoded, we now pass the current flushId value, if it changes when the task runs, drop the message.

We also add some minimal logging.

Added tests.

* LayoutTests/media/media-rvfc-paused-mp4-expected.txt: Added.
* LayoutTests/media/media-rvfc-paused-mp4.html: Added.
* LayoutTests/media/media-rvfc-paused-offscreen-mp4-expected.txt: Added.
* LayoutTests/media/media-rvfc-paused-offscreen-mp4.html: Added.
* LayoutTests/media/media-rvfc-paused-offscreen-webm-expected.txt: Added.
* LayoutTests/media/media-rvfc-paused-offscreen-webm.html: Added.
* LayoutTests/media/media-rvfc-paused-webm-expected.txt: Added.
* LayoutTests/media/media-rvfc-paused-webm.html: Added.
* LayoutTests/media/media-vp8-hiddenframes.html: Remove FIXME and now use rVFC promise as the issue preventing its used was fixed in 286853@main
* LayoutTests/media/utilities.js:
(waitForVideoFrame): Promise was incorrectly resolved with two elements. Can only have one, return an array.
(waitForVideoFrameUntil): Ditto.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm: FlyBy: only perform a readback if we need to and not whenever a frame is available.
(WebCore::MediaPlayerPrivateWebM::setHasAvailableVideoFrame):
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::decodeNextSample):
(WebCore::VideoMediaSampleRenderer::decodedFrameAvailable):
(WebCore::VideoMediaSampleRenderer::maybeQueueFrameForDisplay):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::flush): Cancel timer before purging the compressed queue, to reduce the chance of having a midflight display.
(WebCore::VideoMediaSampleRenderer::notifyHasAvailableVideoFrame):

Canonical link: https://commits.webkit.org/286949@main
